src/shared/auto-update-cache.ts

```diff
@@ -0,0 +1,58 @@
+import { TimeString } from '@types';
+import ms, { StringValue } from 'ms';
+import { timeoutPromise } from './timeouts';
+type Timestamp = number;
+type CacheConstructorParams<Value> = {
+  calculate: () => Promise<Value>;
+  config: AutoUpdateCacheConfig;
+};
+
+export type AutoUpdateCacheConfig = {
+  options: AutoUpdateConfigOptions;
+};
+
+export type AutoUpdateConfigOptions = {
+  useCachedValue: 'always' | { ifUnder: StringValue };
+  nextUpdate: { ifSuccess: StringValue; ifFails: StringValue };
+  timeout?: TimeString;
```

Comment:
Since it's running on the background, I personally wouldn't add this. If we need it at some point, we can add it to whatever is passed to `calculate`, instead of building it into the cache. Like:
```
{
    calculate: timeoutPromise(() => , timeout)
}
```